dace/codegen/targets/framecode.py

```diff
@@ -467,6 +393,325 @@ def dispatch_state(state: SDFGState) -> str:
+        cstate = scope if isinstance(scope, SDFGState) else state
+        # if isinstance(scope, SDFG):
+        #     cstate = None
+        csdfg = scope if isinstance(scope, SDFG) else sdfg
+
+        if desc.storage in dtypes.FPGA_STORAGES:
+            return sdscope.is_devicelevel_fpga(csdfg, cstate, node)
+        elif desc.storage in dtypes.GPU_STORAGES:
+            return sdscope.is_devicelevel_gpu(csdfg, cstate, node)
+
+        # return True
+        return False
+
+    def determine_allocation_lifetime(self, top_sdfg: SDFG):
+        """ 
+        Determines where (at which scope/state/SDFG) each data descriptor
+        will be allocated/deallocated.
+        :param top_sdfg: The top-level SDFG to determine for.
+        """
+        # Gather shared transients
+        shared_transients = {}
+        for sdfg in top_sdfg.all_sdfgs_recursive():
+            shared_transients[sdfg.sdfg_id] = sdfg.shared_transients(
+                check_toplevel=False)
+
+        for sdfg, name, desc in top_sdfg.arrays_recursive():
+            if not desc.transient:
+                continue
+
+            # Possibly confusing control flow below finds the first state
+            # and node of the data descriptor, or continues the
+            # arrays_recursive() loop
+            first_state_instance: int = None
+            first_node_instance: nodes.AccessNode = None
+            last_state_instance: int = None
+            last_node_instance: nodes.AccessNode = None
+            first = True
+            for id, state in enumerate(sdfg.nodes()):
+                for node in state.data_nodes():
+                    if node.data == name:
+                        if first:
+                            first_state_instance = id
+                            first_node_instance = node
+                            first = False
+                        last_state_instance = id
+                        last_node_instance = node
+                        # break
+                else:
+                    continue
+                break
+
+            # Cases
+            if desc.lifetime is dtypes.AllocationLifetime.Persistent:
+                # Persistent memory is allocated in initialization code and
+                # exists in the library state structure
+
+                # If unused, skip
+                if first_node_instance is None:
+                    continue
+
+                # TODO: The change below was done so that test_alloc_persistent
+                # passes (tests/codegen/allocation_lifetime_test.py)
+                definition = desc.as_arg(name=f'__{sdfg.sdfg_id}_{name}') + ';'
+                # definition = desc.as_arg(name=f'{name}') + ';'
+                self.statestruct.append(definition)
+
+                # self.to_allocate[top_sdfg].append(
+                #     (sdfg.sdfg_id, sdfg.node_id(state), node))
+                self.to_allocate[top_sdfg].append(
```

Comment:
Could you please describe (here or where `to_allocate` is defined), what's the meaning of the different tuple fields?

Reply:
Added in lines ~450-460.